### PR TITLE
Harden close helper semantics

### LINKED ISSUE #30: Harden Close semantics across Clients, Env, and RuntimeEnv
The package close helpers currently have inconsistent lifecycle semantics.

- RuntimeEnv.Close() is nil-safe, but Env.Close() is not.
- Clients.Close(), Env.Close(), and RuntimeEnv.Close() are not idempotent.
- Emulator.Close() and omniRuntime.Close() always use context.Background() with no deadline.

These are easy footguns in test code: callers can reasonably combine defer env.Close() with t.Cleanup(env.Close) or call Close() defensively from multiple paths, then get noisy follow-up errors from repeated schema deletion, client close, or container termination.

Expected direction: clarify and align the contract for all close helpers so public pointer receivers are nil-safe where appropriate, lifecycle-owning close paths are idempotent, and shutdown can be bounded instead of always using context.Background().

### PR DESCRIPTION
**Pull request by apstndb**
## Summary
- make close helpers nil-safe and idempotent across Clients, Env, RuntimeEnv, Emulator, LazyRuntime, LazyEmulator, and the Omni runtime
- avoid nil-pointer panics in Clients.Close by turning missing teardown clients into explicit close errors
- add focused regression tests for repeated close calls, nil receivers, and cached close results

## Validation
- go test ./... -run 'Test(ClientsCloseNilSafeAndIdempotent|EnvCloseNilSafeAndIdempotent|EmulatorCloseNilSafeAndIdempotent|RuntimeEnvClose(ZeroValue|CachesFirstResult)|LazyCloseNilReceivers|OmniRuntimeCloseZeroValue)$'
- golangci-lint run
- go test ./... *(fails in this environment because testcontainers cannot create the reaper container: Docker bridge network not found)*

Closes #30

### CODE REVIEW

File: emulator.go [hunk 1]
```diff
@@ -47,8 +50,20 @@ func (e *Emulator) ClientOptions() []option.ClientOption {
...
+// return the result of that first call.
 func (e *Emulator) Close() error {
-	return e.container.Terminate(context.Background())
+	if e == nil {
+		return nil
+	}
+	if e.closed {
+		return e.closeErr
+	}
+	e.closed = true
```
[the remark this answers is not in the document]
Comment (apstndb) [reply emulator.go]: Switched the shared Close idempotency over to a sync.Once-backed closeState so concurrent callers now block on the first close and observe the same cached result. I also added bounded shutdown contexts for the container/control-plane paths and a concurrent regression test in close_test.go. The data client close remains best-effort because spanner.Client.Close does not return an error; admin/container shutdown errors are still surfaced from Close.

File: emulator.go [hunk 2]
```diff
@@ -142,9 +163,24 @@ func (e *Env) Emulator() *Emulator {
...
-		e.Clients.Close(),
-		e.emulator.Close(),
-	)
+	if e == nil {
+		return nil
+	}
+	if e.closed {
+		return e.closeErr
+	}
+	e.closed = true
```
[the remark this answers is not in the document]
Comment (apstndb) [reply emulator.go]: Switched the shared Close idempotency over to a sync.Once-backed closeState so concurrent callers now block on the first close and observe the same cached result. I also added bounded shutdown contexts for the container/control-plane paths and a concurrent regression test in close_test.go. The data client close remains best-effort because spanner.Client.Close does not return an error; admin/container shutdown errors are still surfaced from Close.

File: omni.go [hunk 3]
```diff
@@ -77,7 +80,17 @@ func finalizeManagedOmniClientConfig(config *spanner.ClientConfig, disableBacken
...
 // Close terminates the Omni container.
 func (o *omniRuntime) Close() error {
-	return o.container.Terminate(context.Background())
+	if o == nil {
+		return nil
+	}
+	if o.closed {
+		return o.closeErr
+	}
+	o.closed = true
```
[the remark this answers is not in the document]
Comment (apstndb) [reply omni.go]: Switched the shared Close idempotency over to a sync.Once-backed closeState so concurrent callers now block on the first close and observe the same cached result. I also added bounded shutdown contexts for the container/control-plane paths and a concurrent regression test in close_test.go. The data client close remains best-effort because spanner.Client.Close does not return an error; admin/container shutdown errors are still surfaced from Close.

File: runtime.go [hunk 4]
```diff
@@ -133,10 +136,16 @@ func (e *RuntimeEnv) Runtime() Runtime {
...
+// Close is nil-safe and idempotent. After the first call, subsequent calls
+// return the result of that first call.
 func (e *RuntimeEnv) Close() error {
 	if e == nil {
 		return nil
 	}
+	if e.closed {
+		return e.closeErr
+	}
+	e.closed = true
```
[the remark this answers is not in the document]
Comment (apstndb) [reply runtime.go]: Switched the shared Close idempotency over to a sync.Once-backed closeState so concurrent callers now block on the first close and observe the same cached result. I also added bounded shutdown contexts for the container/control-plane paths and a concurrent regression test in close_test.go. The data client close remains best-effort because spanner.Client.Close does not return an error; admin/container shutdown errors are still surfaced from Close.

File: spanemuboost.go [hunk 5]
```diff
@@ -68,31 +71,53 @@ func (c *Clients) URI() string {
...
+// return the result of that first call.
 func (c *Clients) Close() error {
-	c.Client.Close()
+	if c == nil {
+		return nil
+	}
+	if c.closed {
+		return c.closeErr
+	}
+	c.closed = true
```
[the remark this answers is not in the document]
Comment (apstndb) [reply spanemuboost.go]: Switched the shared Close idempotency over to a sync.Once-backed closeState so concurrent callers now block on the first close and observe the same cached result. I also added bounded shutdown contexts for the container/control-plane paths and a concurrent regression test in close_test.go. The data client close remains best-effort because spanner.Client.Close does not return an error; admin/container shutdown errors are still surfaced from Close.

File: spanemuboost.go [hunk 6]
```diff
@@ -68,31 +71,53 @@ func (c *Clients) URI() string {
...
+
+	if c.Client != nil {
+		c.Client.Close()
+	}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply spanemuboost.go]: Updated the Clients.Close doc comment to match the implementation: the data client is closed first, then fixed-ID resource teardown runs, then the admin clients are closed.

File: spanemuboost.go [hunk 7]
```diff
@@ -68,31 +71,53 @@ func (c *Clients) URI() string {
...
 			Name: c.InstancePath(),
 		}); err != nil {
 			dropErrs = append(dropErrs, fmt.Errorf("delete instance %s: %w", c.InstancePath(), err))
 		}
 	} else if c.dropDatabase {
-		if err := c.DatabaseClient.DropDatabase(ctx, &databasepb.DropDatabaseRequest{
+		if c.DatabaseClient == nil {
+			dropErrs = append(dropErrs, fmt.Errorf("drop database %s: database admin client is nil", c.DatabasePath()))
+		} else if err := c.DatabaseClient.DropDatabase(ctx, &databasepb.DropDatabaseRequest{
 			Database: c.DatabasePath(),
```
[the remark this answers is not in the document]
Comment (apstndb) [reply spanemuboost.go]: Replaced context.Background() in the close path with a bounded close context (30s via newCloseContext) so shutdown is finite in CI and local hangs do not block Close indefinitely.

File: emulator.go [hunk 8]
```diff
@@ -47,8 +50,20 @@ func (e *Emulator) ClientOptions() []option.ClientOption {
...
+	if e.container != nil {
+		e.closeErr = e.container.Terminate(context.Background())
+	}
+	return e.closeErr
```
[the remark this answers is not in the document]
Comment (apstndb) [reply emulator.go]: Replaced context.Background() in the close path with a bounded close context (30s via newCloseContext) so shutdown is finite in CI and local hangs do not block Close indefinitely.

File: omni.go [hunk 9]
```diff
@@ -77,7 +80,17 @@ func finalizeManagedOmniClientConfig(config *spanner.ClientConfig, disableBacken
...
+	if o.container != nil {
+		o.closeErr = o.container.Terminate(context.Background())
+	}
+	return o.closeErr
```
[the remark this answers is not in the document]
Comment (apstndb) [reply omni.go]: Replaced context.Background() in the close path with a bounded close context (30s via newCloseContext) so shutdown is finite in CI and local hangs do not block Close indefinitely.

File: spanemuboost.go [hunk 10]
```diff
@@ -64,35 +66,55 @@ func (c *Clients) URI() string {
...
 		}
-	} else if c.dropDatabase {
-		if err := c.DatabaseClient.DropDatabase(ctx, &databasepb.DropDatabaseRequest{
-			Database: c.DatabasePath(),
-		}); err != nil {
-			dropErrs = append(dropErrs, fmt.Errorf("drop database %s: %w", c.DatabasePath(), err))
+
+		var dropErrs []error
+		ctx, cancel := newCloseContext()
+		defer cancel()
```
[the remark this answers is not in the document]
Comment (apstndb) [reply spanemuboost.go]: Moved newCloseContext() inside the teardown branch so Clients.Close only allocates a bounded context when instance/database cleanup is actually requested. The instance-delete path already skips an explicit database drop, so the close path now follows both parts of the suggestion.

File: emulator.go [hunk 11]
```diff
@@ -13,6 +13,8 @@ import (
 type Emulator struct {
 	container *tcspanner.Container
 	opts      *emulatorOptions
+
+	closeState closeState
```
[the remark this answers is not in the document]
Comment (apstndb) [reply emulator.go]: Moved closeState behind a lazily initialized pointer for exported structs, so Emulator/Env/RuntimeEnv/Clients keep their previous value semantics instead of embedding sync.Once directly. The helper now initializes the shared close state under a package-level mutex, so concurrent Close calls still see the same cached result without introducing a copylock field into exported types. I also simplified Clients.Close to accumulate all returned errors in a single slice.

File: emulator.go [hunk 12]
```diff
@@ -134,6 +151,8 @@ func (le *LazyEmulator) Close() error {
 type Env struct {
 	*Clients
 	emulator *Emulator
+
+	closeState closeState
 }
```
[the remark this answers is not in the document]
Comment (apstndb) [reply emulator.go:158]: Moved closeState behind a lazily initialized pointer for exported structs, so Emulator/Env/RuntimeEnv/Clients keep their previous value semantics instead of embedding sync.Once directly. The helper now initializes the shared close state under a package-level mutex, so concurrent Close calls still see the same cached result without introducing a copylock field into exported types. I also simplified Clients.Close to accumulate all returned errors in a single slice.

File: spanemuboost.go [hunk 13]
```diff
@@ -64,35 +66,57 @@ func (c *Clients) URI() string {
...
+			}
 		}
-	}
 
-	return errors.Join(append(dropErrs,
-		c.DatabaseClient.Close(),
-		c.InstanceClient.Close(),
-	)...)
+		var errs []error
+		errs = append(errs, dropErrs...)
```
[the remark this answers is not in the document]
Comment (apstndb) [reply spanemuboost.go]: Moved closeState behind a lazily initialized pointer for exported structs, so Emulator/Env/RuntimeEnv/Clients keep their previous value semantics instead of embedding sync.Once directly. The helper now initializes the shared close state under a package-level mutex, so concurrent Close calls still see the same cached result without introducing a copylock field into exported types. I also simplified Clients.Close to accumulate all returned errors in a single slice.

File: close.go [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+	if *slot != nil {
+		return *slot
+	}
+	closeStateInitMu.Lock()
+	defer closeStateInitMu.Unlock()
+	if *slot == nil {
+		*slot = &closeState{}
+	}
+	return *slot
+}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply close.go:35]: Dropped the unsynchronized fast path from ensureCloseState so both the read and the first write of the shared close-state pointer now happen under closeStateInitMu. That keeps the exported structs comparable while also removing the race on first Close. Re-ran the full race suite, golangci-lint, and git diff --check after the change.

File: close.go [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+func ensureCloseState(slot **closeState) *closeState {
+	closeStateInitMu.Lock()
+	defer closeStateInitMu.Unlock()
+	if *slot == nil {
+		*slot = &closeState{}
+	}
+	return *slot
+}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply close.go:35]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: spanemuboost.go [hunk 16]
```diff
@@ -43,6 +43,8 @@ type Clients struct {
 
 	dropDatabase bool
 	dropInstance bool
+
+	closeState *closeState
```
[the remark this answers is not in the document]
Comment (apstndb) [reply spanemuboost.go:49]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: spanemuboost.go [hunk 17]
```diff
@@ -64,35 +66,55 @@ func (c *Clients) URI() string {
...
-		// Deleting the instance also removes all databases within it,
-		// so there is no need to drop the database separately.
-		if err := c.InstanceClient.DeleteInstance(ctx, &instancepb.DeleteInstanceRequest{
-			Name: c.InstancePath(),
-		}); err != nil {
-			dropErrs = append(dropErrs, fmt.Errorf("delete instance %s: %w", c.InstancePath(), err))
+	if c == nil {
+		return nil
+	}
+	return ensureCloseState(&c.closeState).close(func() error {
```
[the remark this answers is not in the document]
Comment (apstndb) [reply spanemuboost.go:82]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: emulator.go [hunk 18]
```diff
@@ -13,6 +13,8 @@ import (
 type Emulator struct {
 	container *tcspanner.Container
 	opts      *emulatorOptions
+
+	closeState *closeState
```
[the remark this answers is not in the document]
Comment (apstndb) [reply emulator.go:18]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: emulator.go [hunk 19]
```diff
@@ -47,8 +49,20 @@ func (e *Emulator) ClientOptions() []option.ClientOption {
...
 
 // Close terminates the emulator container.
+// Close is nil-safe and idempotent. After the first call, subsequent calls
+// return the result of that first call.
 func (e *Emulator) Close() error {
-	return e.container.Terminate(context.Background())
+	if e == nil {
+		return nil
+	}
+	return ensureCloseState(&e.closeState).close(func() error {
```
[the remark this answers is not in the document]
Comment (apstndb) [reply emulator.go:59]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: emulator.go [hunk 20]
```diff
@@ -134,6 +151,8 @@ func (le *LazyEmulator) Close() error {
 type Env struct {
 	*Clients
 	emulator *Emulator
+
+	closeState *closeState
```
[the remark this answers is not in the document]
Comment (apstndb) [reply emulator.go:157]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: emulator.go [hunk 21]
```diff
@@ -142,9 +161,20 @@ func (e *Env) Emulator() *Emulator {
...
+// return the result of that first call.
 func (e *Env) Close() error {
-	return errors.Join(
-		e.Clients.Close(),
-		e.emulator.Close(),
-	)
+	if e == nil {
+		return nil
+	}
+	return ensureCloseState(&e.closeState).close(func() error {
```
[the remark this answers is not in the document]
Comment (apstndb) [reply emulator.go:172]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: runtime.go [hunk 22]
```diff
@@ -125,6 +125,8 @@ func isNilRuntimeValue(runtime Runtime) bool {
 type RuntimeEnv struct {
 	*Clients
 	runtime Runtime
+
+	closeState *closeState
```
[the remark this answers is not in the document]
Comment (apstndb) [reply runtime.go:130]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: runtime.go [hunk 23]
```diff
@@ -133,19 +135,22 @@ func (e *RuntimeEnv) Runtime() Runtime {
...
-
-	var errs []error
-	if e.Clients != nil {
-		errs = append(errs, e.Clients.Close())
-	}
-	if e.runtime != nil {
-		errs = append(errs, e.runtime.Close())
-	}
-	return errors.Join(errs...)
+	return ensureCloseState(&e.closeState).close(func() error {
```
[the remark this answers is not in the document]
Comment (apstndb) [reply runtime.go:145]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: close.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (apstndb) [reply close.go:35]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: spanemuboost.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (apstndb) [reply spanemuboost.go:49]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: spanemuboost.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (apstndb) [reply spanemuboost.go:82]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: emulator.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (apstndb) [reply emulator.go:18]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: emulator.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (apstndb) [reply emulator.go:59]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: emulator.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (apstndb) [reply emulator.go:157]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: emulator.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (apstndb) [reply emulator.go:172]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: runtime.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (apstndb) [reply runtime.go:130]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: runtime.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (apstndb) [reply runtime.go:145]: Kept the pointer-backed closeState deliberately. Embedding sync.Once by value in exported structs would make Emulator, Env, RuntimeEnv, and Clients non-comparable and would surface a copylock field as part of their public layout. This update adds inline comments at the helper and field sites so that design constraint is explicit, while still preserving the synchronized first-use initialization path and cached Close result semantics.

File: spanemuboost.go [hunk 24]
```diff
@@ -43,6 +43,9 @@ type Clients struct {
 
 	dropDatabase bool
 	dropInstance bool
+
+	// Pointer-backed to keep exported Clients comparable as a value.
```
[the remark this answers is not in the document]
Comment (apstndb) [reply spanemuboost.go]: Updated the Clients field comment to the actual constraint here: the pointer-backed state avoids embedding a sync.Once copylock in the exported Clients struct layout. The comparability rationale remains on the exported structs where it actually applies.
